rbd: support pvc-pvc clone with different sc & encryption

Rakshith-R opened the pull request:
This commit makes modification so as to allow pvc-pvc clone
with different storageclass having different encryption
configs.

refer: kubernetes-csi/external-provisioner#699

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
rbd: support pvc-pvc clone with different sc & encryption

This commit makes modification so as to allow pvc-pvc clone
with different storageclass having different encryption
configs.
This commit also modifies `copyEncryptionConfig()` to
include a `isEncrypted()` check within the function.

Signed-off-by: Rakshith R <rar@redhat.com>

Madhu-1 commented:
@Rakshith-R add E2E testing with the canary csi-provisioner image to make sure we have covered this case in CI?

Rakshith-R commented:
> @Rakshith-R add E2E testing with the canary csi-provisioner image to make sure we have covered this case in CI?

i did not find canary/latest csi-provisioner image. 
I'll open another pr with e2e and custom built csi-prov image, we can merged that when csi-provisioner release comes out.

Madhu-1 commented:
> > @Rakshith-R add E2E testing with the canary csi-provisioner image to make sure we have covered this case in CI?
> 
> i did not find canary/latest csi-provisioner image. I'll open another pr with e2e and custom built csi-prov image, we can merged that when csi-provisioner release comes out.

`gcr.io/k8s-staging-sig-storage/csi-provisioner:canary` is the canary image.

Rakshith-R committed:
e2e: testcase for pvc-pvc clone with different SC & encryption

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
ci: use canary csi-provisioner image to test different sc clones

This commit is added to use canary csi-provisioner image
to test different sc pvc-pvc cloning feature, which is not
yet present in released versions.
refer:
https://github.com/kubernetes-csi/external-provisioner/pull/699

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
/retest all

Rakshith-R commented:
> /retest all

Jenkins failure
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/upgrade-tests-cephfs/detail/upgrade-tests-cephfs/5353/pipeline/13

Rakshith-R commented on internal/rbd/rbd_util.go:
```diff
@@ -1366,15 +1366,6 @@ func (rv *rbdVolume) cloneRbdImageFromSnapshot(
 		}
 	}()
 
-	if pSnapOpts.isEncrypted() {
```
This is removed and moved to a higher level
Since `cloneRbdImageFromSnapshot` will be called for intermediate temp clone creation too which will not have encryption initialized.

nixpanic replied:
better place this in a new helper function, instead of having multiple (very) similar copies

Rakshith-R replied:
> better place this in a new helper function, instead of having multiple (very) similar copies

@nixpanic does it sound better to move this check within `.copyEncryptionConfig()`?
that way copy is done only when parent image is encrypted.

nixpanic replied:
I guess so, but you'll need to make that clear in the comment documenting the function

Rakshith-R replied:
> I guess so, but you'll need to make that clear in the comment documenting the function

@nixpanic 
Thats done already too https://github.com/ceph/ceph-csi/pull/3066/files#diff-8133ae76a4b0e57aab17837008b76f106b07d4c1cc8631762106e3a79a899247
PTAL

Rakshith-R commented on internal/rbd/clone.go:
```diff
@@ -216,5 +216,12 @@ func (rv *rbdVolume) doSnapClone(ctx context.Context, parentVol *rbdVolume) erro
 		return errClone
 	}
 
+	if parentVol.isEncrypted() {
```
moved as part of https://github.com/ceph/ceph-csi/pull/3066/files#r859717622

Rakshith-R commented on internal/rbd/controllerserver.go:
```diff
@@ -591,6 +592,13 @@ func (cs *ControllerServer) createVolumeFromSnapshot(
 
 	log.DebugLog(ctx, "create volume %s from snapshot %s", rbdVol, rbdSnap)
 
+	if parentVol.isEncrypted() {
```
moved as part of https://github.com/ceph/ceph-csi/pull/3066/files#r859717622

Rakshith-R commented on internal/rbd/clone.go:
```diff
@@ -145,7 +145,7 @@ func (rv *rbdVolume) createCloneFromImage(ctx context.Context, parentVol *rbdVol
 
 	if parentVol.isEncrypted() {
-		err = parentVol.copyEncryptionConfig(&rv.rbdImage, false)
+		err = parentVol.copyEncryptionConfig(&rv.rbdImage, true)
```
Destination pvc in case of restore/clone will have possibly different encryption config already initialized.

Rakshith-R commented on internal/rbd/encryption.go:
```diff
@@ -184,7 +184,7 @@ func (ri *rbdImage) repairEncryptionConfig(dest *rbdImage) error {
 		}
 
-		return ri.copyEncryptionConfig(dest, false)
+		return ri.copyEncryptionConfig(dest, true)
```
Destination pvc in case of restore/clone will have possibly different encryption config already initialized.

Rakshith-R commented on internal/rbd/rbd_journal.go:
```diff
@@ -325,7 +325,7 @@ func (rv *rbdVolume) Exists(ctx context.Context, parentVol *rbdVolume) (bool, er
 
 	if parentVol != nil && parentVol.isEncrypted() {
-		err = parentVol.copyEncryptionConfig(&rv.rbdImage, false)
+		err = parentVol.copyEncryptionConfig(&rv.rbdImage, true)
```
Destination pvc in case of restore/clone will have possibly different encryption config already initialized.

Rakshith-R commented:
/retest ci/centos/upgrade-tests-rbd

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Rakshith-R commented:
> /retest ci/centos/upgrade-tests-rbd
> /retest ci/centos/mini-e2e-helm/k8s-1.22

Failed during rook deployment

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/upgrade-tests-rbd/detail/upgrade-tests-rbd/5367/pipeline#step-77-log-1875
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/2871/pipeline#step-77-log-1763

nixpanic commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -199,8 +199,9 @@ provisioner:
     image:
-      repository: registry.k8s.io/sig-storage/csi-provisioner
-      tag: v3.1.0
+      # TODO: replace with released image version.
```
can you reference the PR or commit that added support for this? that makes it possible to verify what release can be used later on

Rakshith-R replied:
https://github.com/kubernetes-csi/external-provisioner/pull/699

:+1: I'll add this to the commit msg and pr description too.

nixpanic replied:
and most importantly, in the `# TODO: ` comment, please!

nixpanic reviewed (dismissed):
looks good to me, thanks!

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

Rakshith-R commented:
/retest ci/centos/mini-e2e/k8s-1.21

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.23
> /retest ci/centos/mini-e2e/k8s-1.21

cephfs prov deployment failed

nixpanic commented:
**ci/centos/mini-e2e/k8s-1.21** failed to deploy Rook, rebase once #3069 is merged

Rakshith-R commented:
@mergifyio rebase

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1207,14 +1211,12 @@ func (cs *ControllerServer) doSnapshotClone(
-				"config for %q: %v", cloneRbd, cryptErr)
+	cryptErr := parentVol.copyEncryptionConfig(&cloneRbd.rbdImage, false)
+	if cryptErr != nil {
+		log.WarningLog(ctx, "failed copy encryption "+
```
use err here we dont need a new error and correct the return err and log as an error as we are returning error from here?

Rakshith-R replied:
> use err here we dont need a new error and correct the return err and log as an error as we are returning error from here?

Using err will cause a cleanup due to defer func at new line no. 1206.
I don't think we need to cleanup snapshot and clone incase of encryptionconfig copy fails, 
we can pick up from the same spot in a follow-up request
and copy the encryption config here https://github.com/ceph/ceph-csi/blob/43fe5c34cd12546e99657b28e73930d2ba12682a/internal/rbd/controllerserver.go#L1107.

I'll log as log.Error and return cryptErr at line 1219.

Does this sound good @Madhu-1 ?

Madhu-1 replied:
Sounds good but If the encryption copy fails why to keep snapshot and clone?

Rakshith-R replied:
> Sounds good but If the encryption copy fails why to keep snapshot and clone?

yes, :+1:  after some more thinking it makes sense to return err and cleanup everything. 

If cryptErr is intermittent, we should be fine with continuing the process in a follow-up request
But if is repetitive due to wrong configuration or something, the stale snap and clone can never be
deleted automatically since createSnapshot Req can never be successfully completed.

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.22

Failed due to 
```May 4 09:36:35.999: backend images not matching kubernetes resource count,image count 22 kubernetes resource count 21 

backend image

```

Others tests passed PTAL @Madhu-1 @nixpanic

Rakshith-R commented:
@mergifyio rebase

Rakshith-R commented:
@mergifyio rebase

Rakshith-R commented:
@mergifyio rebase
